// thing-it-test/test/dummyActor.js
module.exports = {
    metadata: {
        plugin: "dummyActor",
        label: "DummyActor",
        role: "actor",
        family: "light",
        deviceTypes: ["dummy-device/dummyDevice"],
        services: [{
            id: "on",
            label: "On"
        }, {
            id: "off",
            label: "Off"
        }, {
            id: "blink",
            label: "Blink"
        }],
        state: [{
            id: "light",
            label: "Light",
            type: {
                id: "string"
            }
        }],
        configuration: [{
            label: "Pin",
            id: "pin",
            type: {
                family: "reference",
                id: "digitalInOutPin"
            },
            defaultValue: "12"
        }]
    },
    create: function () {
        return new DummyActor();
    }
};

var q = require('q');

/**
 *
 */
function DummyActor() {
    /**
     *
     */
    DummyActor.prototype.start = function () {
        var deferred = q.defer();

        this.state = {
            light: "off"
        };

        deferred.resolve();

        return deferred.promise;
    };

    /**
     *
     */
    DummyActor.prototype.stop = function () {
    };
    
    /**
     *
     */
    DummyActor.prototype.getState = function () {
        return this.state;
    };

    /**
     *
     */
    DummyActor.prototype.setState = function (state) {
        this.state = state;
    };

    /**
     *
     */
    DummyActor.prototype.on = function () {
        this.state.light = "on";

        this.logInfo(">>> ON");

        this.publishStateChange();
    }

    /**
     *
     */
    DummyActor.prototype.off = function () {
        this.state.light = "off";

        this.logInfo(">>> OFF");

        this.publishStateChange();
    }

    /**
     *
     */
    DummyActor.prototype.toggle = function () {
        if (this.state.light == "off") {
            this.state.light = "on";
        } else {
            this.state.light = "off";
        }

        this.logInfo(">>> TOGGLE");

        this.publishStateChange();
    }

    /**
     *
     */
    DummyActor.prototype.blink = function () {
        this.state.light = "blink";

        this.logInfo(">>> BLINK");

        this.publishStateChange();
    }
};
