// thing-it-test/test/dummySensor.js
module.exports = {
    metadata: {
        plugin: "dummySensor",
        label: "Dummy Sensor",
        role: "sensor",
        family: "rangeSensor",
        deviceTypes: ["dummy-device/dummyDevice"],
        configuration: [{
            label: "Pin",
            id: "pin",
            type: {
                family: "reference",
                id: "analogInPin"
            },
            defaultValue: "A0"
        }, {
            label: "Rate",
            id: "rate",
            type: {
                id: "integer"
            },
            defaultValue: 1000,
            unit: "ms"
        }, {
            label: "Minimum",
            id: "min",
            type: {
                id: "integer"
            },
            defaultValue: 0
        }, {
            label: "Maximum",
            id: "max",
            type: {
                id: "integer"
            },
            defaultValue: 1023
        }]
    },
    create: function () {
        return new DummySensor();
    }
};

/**
 *
 */
function DummySensor() {
    /**
     *
     */
    DummySensor.prototype.start = function () {
        this.interval = setInterval(function () {
            this.value = Math.random();

            this.data(this.value);
        }.bind(this), this.configuration.frequency * 1000);
    };

    /**
     *
     */
    DummySensor.prototype.stop = function () {
        clearInterval(this.interval);
    };
};
